### PR TITLE
Interceptors

### PR DESCRIPTION
**Pull request by mitevs**
## Informations

Type | Status | Migration
---|---|---
Feature | In development | No

****

## Description
Allow for interceptors to wrap service methods. Other components can be injected

## Usage example

```ts
import { Interceptor, IInterceptor, IInterceptorContext, Intercept, Service } from "ts-express-decorators";

@Interceptor()
class MyInterceptor implements IInterceptor {
  aroundInvoke(ctx: IInterceptorContext) {
    //do logic and decide if the call should proceed or not, error can be thrown as well
    // ctx contains { target, method, args } as well so that the aroundInvoke can get more details
    // about the method being intercepted
    ctx.proceed();
  }
}

@Service()
export default MyService {
  @Intercept(MyInterceptor)
  myMethod() {
   ...
  }
}
```

## Todos

- [x] Tests
- [x] Example
- [x] Documentation

### CODE REVIEW

**Comment by mitevs** [comment]
@Romakita I seeded spec for unit testing. I need to execute the decorator from the unit test without the @. Any suggestions? I need to test if the method is correctly wired to use the InjectorService and run the aroundInvoke. This is for the Intercept.ts

File: src/common/interceptors/Intercept.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+          }
+        }, options);
+      }
+
+      return original.apply(this, args);
+    };
+
+    return descriptor;
+  };
+}
```
Comment (Romakita) [inline src/common/interceptors/Intercept.ts]: Intercept.spec.ts:

```typescript
describe('@Intercept', () => {
     class ServiceTest {}
     class InterceptorTest (}
     
    before(() => {
         this.interceptor = {
               aroundInvoke: Sinon.stub()
         };
         this.originalMethod = Sinon.stub();
         this.descriptor = { 
              value: this.originalMethod
         }
         this.getStub = Sinon.stub(InjectoreService, 'get').returns(interceptor);
         this.hasStub = Sinon.stub(InjectoreService, 'has').returns(true);

        this.returnedDescriptor = Intercept(InterceptorTest, {options: 'options'} as any)(ServiceTest, 'method', this.descriptor);
        this.returnedDescriptor.value('arg1');
        this.interceptor.aroundInvoke.getCall(0).args[0].proceed();
    });
     
   after(() => {
         this.getStub.restore();
         this.hasStub.restore();
   });
   
   it('should call InjectorService.has', () => {
        this.hasStub.should.have.been.calledWithExactly(InterceptorTest);
   });
   
   it('should call InjectorService.get', () => {
        this.getStub.should.have.been.calledWithExactly(InterceptorTest);
   });

   it('should call interceptor.aroundInvoke', () => {
        this.interceptor.aroundInvoke.should.have.been.calledWithExactly({
            target: this.descriptor, // not sure
            method: 'method',
            args: ['arg1'],
            proceed: Sinon.match.func
         }, {options:'options'});
   });

   it('should call the original method', () => {
        this.originalMethod.should.have.been.calledWithExactly('arg1');
   });
});
```

Comment (mitevs) [reply src/common/interceptors/Intercept.ts]: Great thanks I will test this now. I haven't been using Sinon too much, and this is very helpful! :)

**Comment by mitevs** [comment]
@Romakita done with the unit tests, please check it and approve. I will proceed to write .md in the docs for the usage and one example. Just let me know where do you put the examples, is it part of the docs file or you have separate location for it. I am asking for convenience. Thanks!

**Comment by Romakita** [comment]
@mitevs The API documentation is based on the code and his comments. 
For a feature like Interceptor, by convention I create a new markdown file under `docs/docs/`. 
You can create your markdown file in the `docs/docs` and I'll create the right link between the sidebar, navbar and your file.

Example: https://github.com/Romakita/ts-express-decorators/blob/master/docs/docs/controllers.emd

**Comment by mitevs** [comment]
@Romakita ok, will do so and ping back. Cheers!

**Comment by Romakita** [comment]
Hi @mitevs,

Tell me when I can merge ? :)

**Comment by mitevs** [comment]
@Romakita Hey, sure.. I will try to finish this during the weekend. I am moving to Munich and a bit in rush this week. I still need to complete the docs before we can merge

Thanks for the understanding, soon to be done! :)

**Comment by mitevs** [comment]
@Romakita  you  can merge the PR, I've put the md file, please check if all looks fine and merge.

Cheers!

**Pull request merged**